docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,210 @@
+
+| Instance Attributes | Type | Description |
+| ------------- | ------------- | ------------- |
+| `niltype` | `typelib` | an immutable runtime representation of the built-in type `nil` |
+| `unknown` | `typelib` | an immutable runtime representation of the built-in type `unknown` |
+| `never` | `typelib` | an immutable runtime representation of the built-in type `never` |
+| `any` | `typelib` | an immutable runtime representation of the built-in type `any` |
+| `boolean` | `typelib` | returns an immutable runtime representation of the built-in type `boolean` |
+| `number` | `typelib` | returns an immutable runtime representation of the built-in type `number` |
+| `string` | `typelib` | returns an immutable runtime representation of the built-in type `string` |
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `issubtypeof(arg: typelib)` | `boolean` | returns true if self is syntactically a subtype or equal to arg in the type hierarchy |
+| `__eq(arg: typelib)` | `boolean` | overrides the == operator to return true if self is syntactically equal to arg in the type hierarchy |
+
+| Static Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getnegation(arg: typelib)` | `typelib` | returns an immutable runtime representation of the negation of the argument; the argument cannot be `istable()`, `ismetatable` or `isfunction()` |
+| `getstringsingleton(arg: string)` | `typelib` | returns an immutable runtime representation of a string singleton type of the argument |
+| `getbooleansingleton(arg: boolean)` | `typelib` | returns an immutable runtime representation of a boolean singleton type of the argument |
+| `getunion(arg: {typelib})` | `typelib` | returns an immutable runtime representation of union type of its argument |
+| `getintersection(arg: {typelib})` | `typelib` | returns an immutable runtime representation of intersection type of its argument |
+| `newtable(props: {[typelib]: typelib}?, indexer: {key: typelib, value: typelib}?)` | `typelib` | returns a mutable runtime representation of a `table` type |
+| `newmetatable(props: {[typelib]: typelib}?, indexer: {key: typelib, value: typelib}?, metatable: typelib?)` | `typelib` | returns a mutable runtime representation of a metatable. `metatable` argument needs to be the same type as `typelib.newtable()` |
+| `newfunction(parameters: {typelib} \| typelib?, returns: {typelib} \| typelib?)` | `typelib` | returns a mutable runtime representation of a `function` type. Calling `newfunction(X)` will by default set `parameters` to `X` |
+| `type(arg: typelib)` | `string` | returns the tag of the argument ("nil", "unknown", "never", "any", "boolean", "number", "string", "boolean singleton", "string singleton", "negation", "union", "intersection", "table", "metatable", "function", "class") |
```

Comment:
Is this written as `typelib.type(ty)`? Because `typeof` covers this use case, you just need to call `lua_newuserdatatagged` and then add a `__type = "the type"` metamethod.

Reply:
Though this leads to a conflict: `typeof(ty) == "string"` could be either the string primitive, or the `string` _type_. `ty:type() == "string"` might be better in this case.

Reply:
I think the original thing I suggested informally was to actually just use `typeof`, but yeah, there's a bit of awkwardness there with `string` and that's why we wanted to do something else. I think `ty:is("string")` feels good though.